### PR TITLE
Enable training on variable-sized batches

### PR DESCRIPTION
**Pull request by hanaol**
### **Motivation**

The current training pipeline assumes either 

- uniformly sized data across all samples in a batch, or
- single-sample processing when dimensions vary

This restricts the ability to work with datasets that naturally contain heterogeneous shapes (e.g., grids at different resolutions), forcing users to manually pad or preprocess data.

### **Solution**

This PR introduces support for training on heterogeneous batches by:

- Implementing a dynamic collate mechanism that groups and prepares samples of varying sizes at runtime.
- Updating the dataloader and relevant training components (e.g., model forward pass or loss function, where applicable) to handle variable-shaped inputs safely.
- Ensuring training remains stable and consistent even when batch elements differ in their spatial dimensions.

These updates allow the training loop to handle variable-sized input data seamlessly, reducing the need for intrusive or manual preprocessing.

### **Notes**

Handling heterogeneous batches may introduce additional overhead compared to fully vectorized uniform-data training. However, enabling this flexibility is valuable for datasets where variable-sized samples are inherent to the problem rather than avoidable.

### CODE REVIEW

File: src/electrai/entrypoints/train.py [hunk 1]
```diff
@@ -11,10 +11,20 @@
...
 from torch.utils.data import DataLoader
+from torch.utils.data._utils.collate import default_collate
 
 torch.backends.cudnn.conv.fp32_precision = "tf32"
 
 
+def collate_fn(batch):
+    try:
+        return default_collate(batch)
+    except Exception:
```
Comment (forklady42) [inline src/electrai/entrypoints/train.py]: This catches all possible exceptions, which is very broad. Some of them could be real errors that should be thrown. Is there a particular class of exceptions that you want to catch here?

File: src/electrai/lightning.py [hunk 2]
```diff
@@ -41,8 +49,16 @@ def training_step(self, batch):
...
+        if isinstance(x, list):
+            losses = []
+            for x_i, y_i in zip(x, y, strict=False):
+                pred = self(x_i.unsqueeze(0))
+                loss = self.loss_fn(pred, y_i.unsqueeze(0))
+                losses.append(loss)
+            loss = torch.stack(losses).mean()
+        else:
+            pred = self(x)
+            loss = self.loss_fn(pred, y)
```
Comment (forklady42) [inline src/electrai/lightning.py]: This logic is a duplicate of that in `training_step` above. It would be good to create a separate function, e.g. `_loss_calculation()` that each of these functions call. That way, next time we update this code, we won't accidentally miss an update to one and cause them to drift.

File: src/electrai/entrypoints/train.py [hunk 3]
```diff
@@ -11,10 +11,20 @@
 from src.electrai.dataloader.registry import get_data
 from src.electrai.lightning import LightningGenerator
 from torch.utils.data import DataLoader
+from torch.utils.data._utils.collate import default_collate
```
Comment (forklady42) [inline src/electrai/entrypoints/train.py]: What's the motivation for importing `default_collate` from `torch.utils.data._utils.collate` rather than `torch.utils.data`?

File: src/electrai/entrypoints/train.py [hunk 4]
```diff
@@ -11,10 +11,20 @@
...
 
 torch.backends.cudnn.conv.fp32_precision = "tf32"
 
 
+def collate_fn(batch):
+    try:
+        return default_collate(batch)
+    except Exception:
+        # Separate and return as lists of tensors
+        x, y = zip(*batch, strict=False)
```
[the remark this answers is not in the document]
Comment (forklady42) [reply src/electrai/entrypoints/train.py]: Agreed that the lengths shouldn't differ, but it's better to be defensive so that future changes or data corruption don't quietly fail.

There's ~three~ two other places in this PR that use zip. Would be good to use `strict=True` for all of them, unless there is a clear reason not to.

I'm less concerned about adding checks to `mp.py`, but it would be fine to do that as well.
